남은 담당 페이지에 Clean Architecture 구조를 적용했어요

seungyooooong opened the pull request:
### 📝 Issue Number
- #32 
- #52 

### 🔧 변경 사항
- RankingView, MainView에 CleanArchitecture를 적용했어요
- MainView에 타이머 기능을 추가했어요

### 🧐 추가 설명
- RankingView는 업데이트가 필요 없는 화면이라고 생각해 viewModel 삭제, API 호출은 MainView로 이동했습니다
- 일정이 바뀌면서 dogether를 먼저 하게돼서 빨리 올려버리겠습니다😅 3/16(일요일)까지 리뷰 부탁드립니다
- 어제 회의때 이야기 나온 수정 부분([참고](https://www.notion.so/1ae483e1832f807eb891c54a02bda2dc))은 따로 수정해서 올리도록 하겠습니다
(폴더링 이야기 나왔던 부분은 아예 일요일 회의때 같이 있을 때 진행하고 포크까지 하는 게 나을 것 같아요)

seungyooooong committed:
feat #52: RankingView 수정

- 업데이트가 필요 없는 화면이라고 생각해 viewModel을 삭제했어요

seungyooooong committed:
feat #52: MainView 수정(1)

seungyooooong committed:
feat #52: MainView 수정(2)

+ GroupInfoView 추가

seungyooooong committed:
feat #52: MainView 수정(3)

seungyooooong committed:
feat #52: MainView 수정(4)

seungyooooong committed:
feat #52: MainView 수정(5)

jieun0330 commented on dogether/Domain/UseCase/MainUseCase.swift:
```diff
@@ -0,0 +1,65 @@
+    func getMainViewStatus() async throws -> MainViewStatus {
+        let response = try await repository.getGroupStatus()
+        let groupStatus = response.status
+        return groupStatus == "READY" ? .beforeStart : .emptyList
```
🌱
`"READY"` 부분도 `enum`으로 분리하면 유지보수성이 좋아질 것 같습니다!

seungyooooong replied:
앗 저 API는 데모데이 전날에 성욱님이랑 저랑 작업하면서 필요에 의해 급하게 만든거라 아직 API 명세 자체가 안나와있어요 ㅎㅎ...

아마 그룹 종료 시나리오가 안정해져있다보니까 (원래 처음 기획대로 메인화면에서 종료된 그룹의 정보를 보여준다면 "FINISHED"같은 응답도 있을텐데 지금은 메인화면에서 종료된 그룹의 정보를 안보여주니까 아마 없겠네요..?) 백에서도 어떤 종류의 응답을 내려주는지 정해지지 않았을거에요

추후에 백엔드쪽 개선 보고 API 명세가 나오면 MainViewStatus와 합치면서 따라가도록 하겠습니다!

jieun0330 commented on dogether/Presentation/Common/GroupInfoView.swift:
```diff
@@ -0,0 +1,119 @@
+        return label
+    }
+    
+    final class InfoLabel: UILabel {
```
🚩
`GroupInfoView` 내부에 `final class InfoLabel`을 정의하신 이유가 궁금합니다.

seungyooooong replied:
1. lineHeight를 유지하려는 목적으로 attributedText를 사용함
2. attributedText.string이 get only-property여서 수정하는 방법을 못찾겠음
3. attributedText 자체를 새로 선언해서 바꾸는 게 나을지 class로 새로 선언해서 updateText라는 함수를 호출하는 게 나을지 고민함
```
        durationInfoLabel.attributedText = NSAttributedString(
            string: "\(groupInfo.duration)일",
            attributes: Fonts.getAttributes(for: Fonts.body1S, textAlignment: .left)
        ) // 1안
        durationInfoLabel.updateText("\(groupInfo.duration)일") // 2안
```
4. GroupInfoView의 관점에서 변수 선언 부분 (개인적으로 74라인까지를 변수 선언 부분이라고 생각해요)보다는 그 아래 함수들의 가독성이 중요하다고 판단 (= updateUI 함수의 가독성 측면에서 updateText라는 함수를 호출하는 게 직관적이고 가독성이 높다고 생각했어요)

이런 의식의 흐름을 타고 가다보니 저렇게 됐어요

근데 지금보니 1안이 더 나은 것 같기도.. 하네요 지은님이 보시기엔 어떤가요?

jieun0330 replied:
저도 1안이 더 직관적인 것 같아요!

jieun0330 commented on dogether/Presentation/Common/GroupInfoView.swift:
```diff
@@ -0,0 +1,119 @@
+            )
+        }
+        
+        private func setUI() {
```
🚩
`updateUI()`와 `setUI()`를 분리한 이유가 궁금합니다

seungyooooong replied:
GroupInfoView는 MainView에서 쓰일 때 날짜같은 데이터가 바뀔 수 있는 컴포넌트라고 생각해요 (D-day같은 경우 날짜가 지나면 바뀌어야 하니까요)

그래서 update로직을 구분하지 않으면 처음 MainView가 생성될 때를 기준으로 데이터를 세팅하고나서 날짜가 지나고 메인 화면을 refresh 하더라도 이전 날짜를 기준으로 D-day가 남아있는 이슈가 예상됐어요

그래서 update로직을 구분했는데, 하다보니 메인화면에서 데이터를 가져오기 전 디폴트 값 (비어있는 GroupInfo)에서 화면을 그리고 -> API를 통해 그룹 정보를 받아오고 -> 그걸 뷰에 다시 적용시키는 구조가 맞다고 생각이 들어서

D-Day부분만이 아니라 API를 통해 가져오는 정보 모두를 update하도록 updateUI 함수를 만들게 됐어요

jieun0330 replied:
`setUI()`에서 `updateUI()`를 호출하는 방식이 초기 UI 설정과 업데이트 로직을 완전히 분리한게 아닌 것 같아 여쭤봤습니다! 뭔가 더 명확히 분리할 수 있는 방법이 있는지 같이 고민해봐요!

Little-tale commented on dogether/Data/DataSources/TodosDataSource.swift:
싱글톤 패턴이 간단한 장점이 있겠지만 
Swift 에선 안티 패턴으로 정해져 있습니다.

모든 객체나 구조체가 접근이 가능해지면
프로젝트가 커질수록 해당 클래스의 책임이 너무 커지기 때문이죠.

seungyooooong replied:
DataSource는 DI Container같은 걸 추가해 관리하는 게 좋을까요?
(+ repository나 useCase도 같이 관리하는 게 좋을까요?)

Manager라고 만들어 둔 것들이 싱글톤 패턴인 게 많은데 그 녀석들은 어떻게 수정하는 게 좋다고 보시나요? (ex. UserDefaultsManager, S3Manager, PushNoticeManager ...)

Little-tale commented on dogether/Domain/Interface/MainInterface.swift:
Interface 라고 하셔도 좋지만
iOS에선 이같은 구성을 Delegate로 부른답니다.

seungyooooong replied:
음 근데 여기서 Interface는 추상화의 의미를 갖는데

iOS에서 말하는 Delegate는 뭔가 행동을 전달하는..? 그런 역할이지 않나요?
delegate라는 게 '전달'이나 '위임'같은 의미라고 알고있는데

재형님 말씀의 의도를 조금 더 자세히 듣고싶습니다

Little-tale commented on dogether/Presentation/View/RankingViewController.swift:
```diff
@@ -10,12 +10,14 @@ import UIKit
+    var rankings: [RankingModel]?
     
     private let navigationHeader = NavigationHeader(title: "순위")
+    
```
각 뷰들의 주석을 달아주시는 습관이 필요해 보여요

당장은 모두가 보일수 있지만 1년뒤에 다시 본다고 생각해보시면서 달아주셔야 할것 같아요

Little-tale commented on dogether/Presentation/ViewModel/MainViewModel.swift:
```diff
@@ -28,59 +29,126 @@ final class MainViewModel {
+            todoList = try await mainUseCase.getTodoList(dateOffset: dateOffset, currentFilter: currentFilter)
+            mainViewStatus = currentFilter == .all && todoList.isEmpty ? .emptyList : .todoList
+            isBlockPanGesture = await todoListHeight < Int(UIScreen.main.bounds.height - (SheetStatus.normal.offset + 140 + 48))
+            await MainActor.run { updateList() }
```
흠 @escaping으로 반환 할거면 Task 로 비동기 처리할 필요가 있었을지 의문입니다.

seungyooooong replied:
말씀해주신 부분에 대해서 생각을 좀 해봤는데 이 부분에서 좀 막히는 것 같아요 비동기 처리에 대해서 Task말고 추천해주시는 방법이 따로 있을까요?

seungyooooong replied:
Task나 escaping의 사용에 대해서 좀 자세히 의견을 나누고 싶은데 일요일은 시간이 안되신다고 해서..

혹시 화요일 정기회의 전에 시간 괜찮으시면 조금 일찍 와주실 수 있을까요?

궁금한 것들 좀 여쭤보는 정도라 많이는 아니고 20분.. 정도..? 화요일 7시 40분 괜찮으시다면 부탁드립니다!

Little-tale replied:
이스케이핑 접근은 좋으나
단순 이스케이핑은 애매해요

Swiftconcurrency 를 쓰는 이유중에 하나는 
데이터 래이스를 막고자 하는 방향이 존재합니다 
참고 @Sendable  

흠 예시적으로 해당하는 것같이 저는 설계를 하는데 
도움이 되실까 싶어요

static func run(priority: TaskPriority? = nil, task: @escaping @Sendable (_ send: @escaping (Action) async -> Void) async -> Void) -> Self {
        Somthings
    }

Little-tale commented on dogether/Presentation/Common/DogetherTodoItem.swift:
```diff
@@ -10,24 +10,28 @@ import UIKit
+    private(set) var action: () async -> Void
     private(set) var todo: TodoInfo
     
     init(action: @escaping () -> Void, todo: TodoInfo) {
```
@escaping 클로저를 앞에다가 하시는 부분이 좀 의문입니다.

이게 SwiftUI Button 을 모방하시려 하셨난 문뜩 떠올르긴 하였는데요
그 버튼도 내부를 보면 Action, Label 모두 클로저 형태입니다.

뒤에다가 하게되면 
예를들어. 
DogetherTodoItem( 무언가 ). {
   
}
같은 현태로 해줄수가 있는데 이부분은 아쉬운 부분 같아요

Little-tale commented on dogether/Presentation/Common/DogetherTodoItem.swift:
```diff
@@ -10,24 +10,28 @@ import UIKit
 import SnapKit
 
 final class DogetherTodoItem: UIButton {
```
버튼인데 이름을 Item 으로 끝난 부분이 너무 난해 한 것 같습니다.

seungyooooong replied:
TodoList에서 사용하는 Item이기 때문에 이렇게 된 것 같은데..

기획적으로 Item 안에 button 형태를 갖는 buttonLabel이 따로 있지만 탭 했을 때 액션은 Item 자체를 모두 동일한 액션으로 가져가기로 정했거든요

이럴 경우에 한 컴포넌트가 List의 Item 역할을 하지만 Button의 특징(탭했을 때 액션을 가진다는)을 갖는데 네이밍 규칙같은 게 있을까요?

Little-tale replied:
각 UI에 맞는 이름을 마지막에 적어주는 편이에요
UIButton ~> 버튼이름Button
UILabel ~> 라벨이름Label ...etc

Little-tale replied:
UI는 단순 컨포넌트라고 전 생각하기 때문에 밖에서 액션을 연결하거나 하는 방향이라

일단은 네이밍은 생각해보시면 좋을것 같아요

후에 파일이 많아지면 UI를 찾을때....를 생각해 보신다면..!

Little-tale commented on dogether/Presentation/Common/FilterButton.swift:
```diff
@@ -64,19 +69,16 @@ final class FilterButton: UIButton {
         }
     }
     
+    func setAction(_ action: @escaping (FilterTypes) -> Void) {
```
액션을 갈아 끼우는 작업을 해당 버튼이 과연 
알아야 했을까가 의문입니다.

버튼은 눌렀음 같은 행위만 알려줄 뿐이지 어떤 액션을 가질건지를 정의한건
흠.. 애매하다고 생각을 합니다.

Little-tale reviewed (approved):
전체적으로 보았을때
Task 가 과연 필요했을까 의문인 코드들이 보여요

그리고 과연 뷰모델에게 hideFloating 이라는 함수를 전달하여야 했을까도 고민해 보셔도
좋을것 같습니다.

사람마다 MVVM ViewModel의 관점은 다르지만 
저는 UI의 상태를 알필요는 없다는 파여성 
Ex) Import UIKit or UIKit.UIImage ... etc 
View Action ( hide ) -> VM ( state.toggle() ) -> VIew ( deinit ) 해당 형태에서
무엇이 비즈니스 로직이였을까 고민해 보셔도 좋을것 같아요


Task 를 여러개 선언하면 
해당 테스크의 부모는 각자가 되며, 
priority 도 정해주시지 않으셨어요!
가장 기본값인 medium 상태가 되는데
모두가 medium 이라면 어떻게 될지 고민해 보셔도 좋을것 같습니다.

중앙 Task 가 없고 각자가 Task 일때
Task.cancel 관리를 어떻게 해야할지 난해해 질 거에요
고민해 보셨으면 합니다.

seungyooooong committed:
feat #52: 리뷰 내용 반영

1. UIButton이 action을 가져야 하는 상황 분류 -> Button의 내부 변수가 action에 영향을 주지 않을 때 action 삭제 (ex. DogetherButton)
2. GroupInfoView 내부 class InfoLabel 삭제
3. DogetherTodoItem -> TodoListItemButton 네이밍 수정
+ MainViewController 내부 updateList 함수에 임시 로직 추가 (추후 개선)

seungyooooong committed:
feat #52: 리뷰 내용 반영

1. UIButton의 내부 변수가 action에 영향을 줄 때 클로저를 맨 뒤로 수정, 기본 action(NA) 지정
